### PR TITLE
a8n: Be liberal in what to accept as Refs in BitbucketServerSource

### LINKED ISSUE #6895: a8n: Creating pull requests on Bitbucket Server fails due to wrong refname
Right now creating PRs on Bitbucket Server fails because of this error message:

```
 Bitbucket API HTTP error: code=404 url="https://bitbucket.sgdev.org/rest/api/1.0/projects/SOUR/repos/automation-testing/pull-requests" body="{"errors":[{"context":null,"message":"Repository \"automation-testing\" of project with key \"SOUR\" has no branch \"refs/heads/refs/heads/master\"","exceptionName":"com.atlassian.bitbucket.commit.NoSuchCommitException"}]}"
```

Note the `refs/heads/refs/heads/master`.

That's probably due to my changes in https://github.com/sourcegraph/sourcegraph/commit/49b07982a9fbce4c7864dfa0b95340d8f02ff0fb

### PR DESCRIPTION
**Pull request by mrnugget**
This fixes #6895 by changing the BitbucketServerSource to check whether the passed in Ref has a `refs/heads/` prefix or not. If it doesn't, it prefixes it.

This was caused by the change in 2dfeb7b as part of #6682 in which we persist the base ref of a repository on the `CampaignJob`. I wrongly assumed that the ref returned by `gitserver` is only the name, i.e. `master` and not `refs/heads/master`.

I changed the relevant tests to reflect that more.

### CODE REVIEW

**Comment by mrnugget** [comment]
> minor: this would be nice as a tiny helper function but I can't think of a nice name for it.

I'm also unclear on the naming in general. Is `refs/heads/master` a _ref_ or a _refspec_ or a _ref name_? Or is `master` the _ref name`?

**Comment by keegancsmith** [comment]
> > minor: this would be nice as a tiny helper function but I can't think of a nice name for it.
> 
> I'm also unclear on the naming in general. Is `refs/heads/master` a _ref_ or a _refspec_ or a _ref name_? Or is `master` the _ref name`?

https://git-scm.com/docs/gitglossary#Documentation/gitglossary.txt-aiddefrefaref

So `refs/...` is a _ref_ (+ some special ones like `HEAD`). Some commands accept abbreviated (ie without the `refs/head/`) for convenience, and as such so does Sourcegraph. A _refspec_ is a mapping. I'm not sure what a _ref name_ is, but I'd suspect it is the abbreviated form.

**Comment by mrnugget** [comment]
Thanks! That's a handy glossary. I'll try to clean up what we already build to be consistent in its naming.

File: cmd/repo-updater/repos/bitbucketserver.go [hunk 1]
```diff
@@ -111,11 +111,21 @@ func (s BitbucketServerSource) CreateChangeset(ctx context.Context, c *Changeset
...
+		pr.ToRef.ID = c.BaseRefName
+	} else {
+		pr.ToRef.ID = fmt.Sprintf("refs/heads/%s", c.BaseRefName)
+	}
 
 	pr.FromRef.Repository.Slug = repo.Slug
 	pr.FromRef.Repository.Project.Key = repo.Project.Key
-	pr.FromRef.ID = fmt.Sprintf("refs/heads/%s", c.HeadRefName)
+
+	if strings.HasPrefix(c.HeadRefName, "refs/heads/") {
```
Comment (tsenart) [inline cmd/repo-updater/repos/bitbucketserver.go]: `"refs/heads/" + strings.TrimPrefix(c.HeadRefName, "refs/heads/")` ?

**Pull request merged**